packages/menu/src/useMenu.ts

```diff
@@ -0,0 +1,644 @@
+        }
+      }
+    },
+    [closeMenu, triggerRef]
+  );
+
+  const handleMenuBlur = useCallback(
+    event => {
+      const path = event.composedPath();
+
+      if (!path.includes(menuRef.current) && !path.includes(triggerRef.current)) {
+        closeMenu(StateChangeTypes.MenuBlur);
+      }
+    },
+    [closeMenu, menuRef, triggerRef]
+  );
+
+  const handleMenuMouseLeave = useCallback(() => {
+    dispatch({
+      type: StateChangeTypes.MenuMouseLeave,
+      payload: {}
+    });
+
+    onChange({ type: StateChangeTypes.MenuMouseLeave });
+  }, [onChange]);
+
+  /**
+   * Effects
+   */
+
+  /**
+   * Sets a state value to ensure 1 render happens after
+   * opening so a menu item can be focused.
+   */
+  useEffect(() => {
+    setMenuVisible(controlledIsExpanded);
+  }, [controlledIsExpanded, menuRef, environment]);
+
+  /**
+   * Respond to clicks outside the  open menu
+   */
+  useEffect(() => {
+    const doc = environment || document;
+
+    if (controlledIsExpanded) {
+      doc.addEventListener('click', handleMenuBlur, true);
+    } else if (!controlledIsExpanded) {
+      doc.removeEventListener('click', handleMenuBlur, true);
+    }
+
+    return () => {
+      doc.removeEventListener('click', handleMenuBlur, true);
+    };
+  }, [controlledIsExpanded, handleMenuBlur, environment]);
+
+  /**
+   * Return focus to trigger when menu closes
+   */
+  useEffect(() => {
+    if (controlledIsExpanded && !returnFocusRef.current) {
+      returnFocusRef.current = triggerRef.current;
+    }
+
+    if (!controlledIsExpanded && returnFocusRef.current) {
+      setTimeout(() => {
+        returnFocusRef.current!.focus();
+        returnFocusRef.current = null;
+      });
+    }
+  }, [triggerRef, controlledIsExpanded]);
+
+  /**
+   * Focus initial item when menu opens
+   */
+  useEffect(() => {
+    if (state.focusOnOpen && menuVisible && controlledFocusedValue && controlledIsExpanded) {
+      const ref = itemRefs[controlledFocusedValue]?.current;
+
+      ref && ref.focus();
+    }
+  }, [menuVisible, itemRefs, controlledFocusedValue, state.focusOnOpen, controlledIsExpanded]);
+
+  /**
+   * Prop getters
+   */
+
+  const getTriggerProps = useCallback<IUseMenuReturnValue<T, L>['getTriggerProps']>(
+    ({ onClick, onKeyDown, type = 'button', role = 'button', disabled, ...other } = {}) => ({
+      'data-garden-container-id': 'containers.menu.trigger',
+      'data-garden-container-version': PACKAGE_VERSION,
+      ref: triggerRef,
+      id: triggerId,
+      'aria-expanded': state.isExpanded,
+      'aria-haspopup': true,
+      disabled,
+      tabIndex: disabled ? -1 : 0,
+      type: type === null ? undefined : type,
+      role: role === null ? undefined : role,
+      ...other,
+      onKeyDown: composeEventHandlers(onKeyDown, handleTriggerKeyDown),
+      onClick: composeEventHandlers(onClick, handleTriggerClick)
+    }),
+    [triggerRef, state.isExpanded, handleTriggerClick, handleTriggerKeyDown, triggerId]
```

Comment:
Definitely, it'll move to `useCallback`.

Reply:
Ha – I was more on the side of bringing the `handleTrigger...` functions internal to `getTriggerProps` in order to avoid excessive callback declarations. Closer to the style of `useCombobox` getters like https://github.com/zendeskgarden/react-containers/blob/18a8cb4c3d4d279147e31ef08fdee6792ff2d6c8/packages/combobox/src/useCombobox.ts#L432